searchlib/src/vespa/searchlib/attribute/postinglistsearchcontext.hpp

```diff
@@ -244,25 +244,46 @@ PostingListSearchContextT<DataT>::approximateHits() const
 
 template <typename DataT>
 void
-PostingListSearchContextT<DataT>::applyRangeLimit(int rangeLimit)
+PostingListSearchContextT<DataT>::applyRangeLimit(long rangeLimit)
 {
     if (rangeLimit > 0) {
         DictionaryConstIterator middle = _lowerDictItr;
-        for (int n(0); (n < rangeLimit) && (middle != _upperDictItr); ++middle) {
+        long n = 0;
+        size_t count = 0;
+        for (; (n < rangeLimit) && (count < max_posting_lists_to_count) && (middle != _upperDictItr); ++middle, count++) {
             n += _postingList.frozenSize(middle.getData().load_acquire());
         }
-        _upperDictItr = middle;
-        _uniqueValues = _upperDictItr - _lowerDictItr;
+        if (n >= rangeLimit) {
+            _upperDictItr = middle;
+        } else {
+            size_t offset = ((rangeLimit - n) * count)/n;
+            middle += offset;
+            if (middle.valid()) {
+                _upperDictItr = middle;
+            }
+        }
+
     } else if ((rangeLimit < 0) && (_lowerDictItr != _upperDictItr)) {
         rangeLimit = -rangeLimit;
         DictionaryConstIterator middle = _upperDictItr;
-        for (int n(0); (n < rangeLimit) && (middle != _lowerDictItr); ) {
+        long n = 0;
+        size_t count = 0;
+        for (; (n < rangeLimit) && (count < max_posting_lists_to_count) && (middle != _lowerDictItr); ) {
```

Comment:
I think `count` must be incremented as well?

Reply:
Correct, the tests told me the same.